crates/wasmparser/src/validator/component_types.rs

```diff
@@ -34,56 +35,174 @@ const MAX_FLAT_FUNC_RESULTS: usize = 1;
 /// The maximum lowered types, including a possible type for a return pointer parameter.
 const MAX_LOWERED_TYPES: usize = MAX_FLAT_FUNC_PARAMS + 1;
 
+/// Error used when we are in the process of lowering a component signature to a
+/// core signature and we need to keep retry lowering in a different mode
+/// (e.g. flat vs indirect in linear memory) only for certain kinds of errors.
+enum LowerError {
+    OutOfCapacity,
+    TypeMismatch(BinaryReaderError),
+    Other(BinaryReaderError),
+}
+
+impl From<BinaryReaderError> for LowerError {
+    fn from(v: BinaryReaderError) -> Self {
+        Self::Other(v)
+    }
+}
+
+impl LowerError {
+    #[track_caller]
+    fn unwrap_binary_reader_error(self) -> BinaryReaderError {
+        match self {
+            LowerError::TypeMismatch(e) | LowerError::Other(e) => e,
+            LowerError::OutOfCapacity => panic!(),
+        }
+    }
+}
+
+type LowerResult<T = ()> = Result<T, LowerError>;
+
+enum Variance {
+    Covariant,
+    Contravariant,
+}
+
+/// A list of expected types and how to validate them.
+struct ExpectedLoweredTypes<'a> {
+    types: &'a TypeList,
+    offset: usize,
+
+    /// The expected types themselves.
+    expected: smallvec::SmallVec<[ValType; MAX_LOWERED_TYPES]>,
+
+    /// The index of the next type to check.
+    index: usize,
+
+    /// The description of this type list (for example "parameters") for use in
+    /// error messages.
+    desc: &'static str,
+
+    /// The kind of variance that should be used when type checking the actual
+    /// vs expected type.
+    variance: Variance,
+}
+
+impl<'a> ExpectedLoweredTypes<'a> {
+    fn new(
+        types: &'a TypeList,
+        desc: &'static str,
+        variance: Variance,
+        offset: usize,
+        expected: &[ValType],
+    ) -> Self {
+        Self {
+            types,
+            expected: expected.iter().copied().collect(),
+            index: 0,
+            desc,
+            variance,
+            offset,
+        }
+    }
+
+    fn reset(&mut self) {
+        self.index = 0;
+    }
+
+    fn peek(&self) -> Option<ValType> {
+        self.expected.get(self.index).copied()
+    }
+
+    fn match_ty(&mut self, actual: ValType) -> LowerResult {
+        let expected = match self.peek() {
+            Some(e) => e,
+            // Arity mismatch, but ignore it for now. The arity check at the
+            // end of `ComponentFuncType::lower` will catch this and can
+            // give a more descriptive error than we can here.
+            None => return Ok(()),
```

Comment:
Do you think it might make sense to have a `self.arity_mismatch` which is debug-assert'd to be false right before we leave without an error higher up the stacsk?

Reply:
I'm not sure I am following this suggestion